File: src/main/java/no/skatteetaten/aurora/databasehotel/service/internal/SchemaLabelMatcher.java
package no.skatteetaten.aurora.databasehotel.service.internal;

import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.apache.commons.lang3.StringUtils;

import com.google.common.base.Verify;

import no.skatteetaten.aurora.databasehotel.domain.DatabaseSchema;

public class SchemaLabelMatcher {

    public static Set<DatabaseSchema> findAllMatchingSchemas(Set<DatabaseSchema> schemas, Map<String, String> labels) {

        Verify.verifyNotNull(schemas, "Schemas cannot be null");
        Verify.verifyNotNull(labels, "Labels cannot be null");

        return schemas.stream().filter(schema -> matchesAll(schema, labels)).collect(Collectors.toSet());
    }

    public static boolean matchesAll(DatabaseSchema schema, Map<String, String> labels) {

        Verify.verifyNotNull(schema, "Schema cannot be null");
        Verify.verifyNotNull(labels, "Labels cannot be null");

        boolean matches = true;
        Map<String, String> schemaLabels = schema.getLabels();
        for (Map.Entry<String, String> label : labels.entrySet()) {
            String valueToMatch = label.getValue();
            String schemaLabelValue = schemaLabels.get(label.getKey());
            if (!StringUtils.equals(valueToMatch, schemaLabelValue)) {
                matches = false;
                break;
            }
        }
        return matches;
    }
}

File: src/main/kotlin/no/skatteetaten/aurora/databasehotel/service/internal/SchemaLabelMatcher.kt
package no.skatteetaten.aurora.databasehotel.service.internal

import no.skatteetaten.aurora.databasehotel.domain.DatabaseSchema

object SchemaLabelMatcher {

    fun findAllMatchingSchemas(schemas: Set<DatabaseSchema>, labels: Map<String, String?>): Set<DatabaseSchema> =
        schemas.filter { schema -> matchesAll(schema, labels) }.toSet()

    internal fun matchesAll(schema: DatabaseSchema, labels: Map<String, String?>?): Boolean =
        labels?.all { (key, valueToMatch) -> valueToMatch == schema.labels[key] } ?: true
}
